### PR TITLE
Jp employee detail

### LINKED ISSUE #4: HR should be able to view employee details
## Acceptance Criteria

**Given** a user is viewing the employee list
**When** the user clicks on an individual employee
**Then** the user should be shown a detail view of that employee, and it must contain the following information

1. First name and last name
1. Department
1. Currently assigned computer
1. Training programs they have attended, or plan on attending

## Testing Criteria

* Your test suite must verify that the content of the response has all of the expected values of the employee - name, department, computer, and training programs.

### PR DESCRIPTION
**Pull request by jeremiah3643**
## Link to Ticket
Closes  #4 

## Description of Proposed Changes
- Created mvc for employee details view
- Created integration test for Employee details View

## Steps to Test

Outline the steps to test

```sh
git fetch --all
git checkout JPEmployeeDetail
run the test for Get_DetailContentVerified
verify in browser that all fields are generated for specific User.
On webpage click on employees, then details(for a specific employee)
```

## Impacted Areas in Application

List general components of the application that this PR will affect:

*New MVC and Test. Will not affect other MVCs or Models or Controllers.

## Mentions @username

@khardt
@figamus
@jeremiah3643
@ricky-bruner
@askingalot

## Definition of Done

1. The project must be fully documented. This includes the following:
    1. Complete README that documents the steps to install the code, how to install any dependencies, or system configuration needed.
    1. Every class must be documented with purpose, author, and methods.
    1. Every method must be documented with purpose and argument list - which itself must contain a short purpose for each argument.
1. The project must be able to run fully, and without errors, on each teammate's system.
1. Fulfills every requirement.
Every line of code has been peer reviewed.
1. For projects that require unit testing, core functionality must be identified and have at least one test for each.

### CODE REVIEW

File: BangazonWorkforce/BangazonWorkforce/Controllers/TrainingProgramController.cs [hunk 1]
```diff
@@ -0,0 +1,237 @@
+﻿/*
```
Comment (askingalot) [inline BangazonWorkforce/BangazonWorkforce/Controllers/TrainingProgramController.cs]: Why do you need this controller?

File: BangazonWorkforce/BangazonWorkforce.IntegrationTests/EmployeeTests.cs [hunk 2]
```diff
@@ -239,7 +239,65 @@ private async Task<List<Department>> GetAllDepartments()
...
+
+         /*Author: Jeremiah Pritchard
+          * 
+          * Purpose:Using a specific user to make sure fields are populated.
+          * */
+         [Fact]
+         public async Task Get_EmployeeDetailVerification()
+         {
+         //Arrange
+         string url = "/employee/details/1";
```
Comment (askingalot) [inline BangazonWorkforce/BangazonWorkforce.IntegrationTests/EmployeeTests.cs]: format code.

File: BangazonWorkforce/BangazonWorkforce/Models/ViewModels/EmployeeDetailModel.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        public Department Department { get; set; }
+        public List<string> TrainingPrograms = new List<string>();
+        public List<string> Computers = new List<string>();
+        public List<string> ComputerManufacturer = new List<string>();
+
+        public EmployeeDetailModel(IConfiguration config, int Id, Employee employee)
+        {
+            using (IDbConnection conn = new SqlConnection(config.GetConnectionString("DefaultConnection")))
+            {
+                IEnumerable<TrainingProgram> trainingPrograms = conn.Query<TrainingProgram>($@"
```
Comment (askingalot) [inline BangazonWorkforce/BangazonWorkforce/Models/ViewModels/EmployeeDetailModel.cs]: format code

File: BangazonWorkforce/BangazonWorkforce/Views/Employee/Details.cshtml [hunk 4]
```diff
@@ -36,14 +35,30 @@
...
         </dt>
+        <dd>
+            @Html.DisplayFor(model => model.Computers)
+            @Html.DisplayFor(model => model.ComputerManufacturer)
+        </dd>
+        <dt>
+            @Html.DisplayNameFor(model => model.TrainingPrograms)
+        </dt>
+        <dd>
+            @Html.DisplayFor(model => model.TrainingPrograms)
```
Comment (askingalot) [inline BangazonWorkforce/BangazonWorkforce/Views/Employee/Details.cshtml]: Where is the `@foreach`?

**Review by askingalot — CHANGES_REQUESTED** [review]

**Review by ricky-bruner — CHANGES_REQUESTED** [review]
1. Not sure why there is a Training Programs Controller being used for this instead of private async calls in the bottom on the employees controller. Training Programs Controller isn't called for until much later in the priorities list.

2. Please change the name of the view model file to `EmployeeDetailsViewModel.cs` to mirror the naming convention of the other files.

3. Please move Get_EmployeeDetailVerification() in `EmployeeTest.cs` to be above the private async call for better orgainization and format the code.

4. Something wonky is going on with the connection string in the config file that barring testing, but I dont think this is an issue with your code. Seems like something isnt importing properly.

File: BangazonWorkforce/BangazonWorkforce/Controllers/TrainingProgramController.cs (hunk 1 above)
Comment (jeremiah3643) [reply BangazonWorkforce/BangazonWorkforce/Controllers/TrainingProgramController.cs]: Controller removed

File: BangazonWorkforce/BangazonWorkforce.IntegrationTests/EmployeeTests.cs (hunk 2 above)
Comment (jeremiah3643) [reply BangazonWorkforce/BangazonWorkforce.IntegrationTests/EmployeeTests.cs]: formatted

File: BangazonWorkforce/BangazonWorkforce/Models/ViewModels/EmployeeDetailModel.cs (hunk 3 above)
Comment (jeremiah3643) [reply BangazonWorkforce/BangazonWorkforce/Models/ViewModels/EmployeeDetailModel.cs]: formatted

File: BangazonWorkforce/BangazonWorkforce/Views/Employee/Details.cshtml (hunk 4 above)
Comment (jeremiah3643) [reply BangazonWorkforce/BangazonWorkforce/Views/Employee/Details.cshtml]: foreach added

**Comment by jeremiah3643** [comment]
All requests have been addressed.

**Review by ricky-bruner — CHANGES_REQUESTED** [review]
```
/*
          Author: Jeremiah Pritchard
          Purpose:Using a specific user to make sure fields are populated.
*/
        [Fact]
        public async Task Get_EmployeeDetailVerification()
        {
            //Arrange
            string url = "/employee/details/1";
            //Act
            HttpResponseMessage response = await _client.GetAsync(url);
            //Assert
            response.EnsureSuccessStatusCode(); // Status Code 200-299
            IHtmlDocument DetailPage = await HtmlHelpers.GetDocumentAsync(response);
            Assert.Contains(
                DetailPage.QuerySelectorAll("dd"),
                dd => dd.TextContent.Contains("Mike"));
            Assert.Contains(
                DetailPage.QuerySelectorAll("dd"),
                dd => dd.TextContent.Contains("Parrish"));
            Assert.Contains(
                DetailPage.QuerySelectorAll("span"),
                span => span.TextContent.Contains("No"));
            Assert.Contains(
                DetailPage.QuerySelectorAll("dd"),
                dd => dd.TextContent.Contains("Navy"));
            Assert.Contains(
                DetailPage.QuerySelectorAll("dd"),
                dd => dd.TextContent.Contains("Surface Tablet"));
            Assert.Contains(
                DetailPage.QuerySelectorAll("dd"),
                dd => dd.TextContent.Contains("You Are the First Line of Defence!"));
        }
```

This test is too specific, in that it will fail if any changes at all are made on Mike.

**Review by ricky-bruner — COMMENTED** [review]
In `EmployeeDetailsViewModel`

You are getting:
```
public List<string> Computers = new List<string>();
public List<string> ComputerManufacturer = new List<string>();
```
From two seperate but nearly identical DB Queries. Would it be DRYer to make one query that gets both the make and manufacturer, and then assign the two bits of data accordingly so that its only one query? Also, Computers should be named something more representative, like ComputerMake. They could also just be strings instead of lists.

File: BangazonWorkforce/bangazon.sql [hunk 5]
```diff
@@ -454,6 +454,24 @@ VALUES
 INSERT INTO EmployeeTraining
 (EmployeeId, TrainingProgramId)
 VALUES
+<<<<<<< HEAD
```
Comment (askingalot) [inline BangazonWorkforce/bangazon.sql]: Looks like you've committed an un-merged file

**Review by ricky-bruner — APPROVED** [review]

**Comment by ricky-bruner** [comment]
Only thing left is possibly comments

**Review by askingalot — CHANGES_REQUESTED** [review]
I'm unable to run the DB script to setup for running the app and tests

File: BangazonWorkforce/bangazon.sql (hunk 5 above)
Comment (ricky-bruner) [reply BangazonWorkforce/bangazon.sql]: >.>

Comment (jeremiah3643) [reply BangazonWorkforce/bangazon.sql]: fixed

File: BangazonWorkforce/bangazon.sql [hunk 6]
```diff
@@ -454,7 +454,23 @@ VALUES
 INSERT INTO EmployeeTraining
```
Comment (askingalot) [inline BangazonWorkforce/bangazon.sql]: This insert statement gives me the following error:

```
Msg 110, Level 15, State 1, Line 454
There are fewer columns in the INSERT statement than values specified in the VALUES clause. The number of values in the VALUES clause must match the number of columns specified in the INSERT statement.
```

**Review by askingalot — CHANGES_REQUESTED** [review]

File: BangazonWorkforce/bangazon.sql (hunk 6 above)
Comment (jeremiah3643) [reply BangazonWorkforce/bangazon.sql]: using original SQL

**Review by askingalot — APPROVED** [review]

**Pull request merged**